### ISSUE TITLE
FLEX decoding...almost - always loses synch

### ISSUE THREAD

**Issue by greedy9** [EliasOenal/multimon-ng#40]
Trying to decode pager transmissions, using rtl_fm and multimon-ng on a Raspberry Pi.  This is what I've assembled from various sources, and it almost works...

rtl_fm -A fast -f 929.620M -M fm -E dc -s 22050  -F 0 | multimon-ng -n -p -t raw -a POCSAG512 -a POCSAG1200 -a POCSAG2400 -a FLEX -i -p -v 3 -f alpha  /dev/stdin

I see messages from the FLEX decoder, mostly these:

FLEX: Synchronisation Lost
FLEX: Locked
FLEX: Synchronisation Lost
FLEX: Locked

but periodically, I get:

FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xd6a0 baud=3200 levels=4 polarity=NEG zero=0.013193 envelope=0.195820 symrate=1601.396648
FLEX: State: FIW
FLEX: Synchronisation Lost
FLEX: State: SYNC1
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=NEG zero=0.017672 envelope=0.191791 symrate=1600.031621
FLEX: State: FIW
FLEX: Phase F Data corruption - Unable to fix errors.
FLEX: Unable to decode FIW, too much data corruption
FLEX: State: SYNC1
FLEX: Synchronisation Lost

and a few of these:

FLEX: State: FIW
FLEX: Phase F Fixed 2 errors @ 0x00084000  (0x7ffef99e -> 0x7ff6b99e)
FLEX: Bad Checksum 0x2

which seems to suggest that multimon-ng is detecting the FLEX format, and decoding some of it, but never completely successfully.  

I've played with the gain, antennas, and so on, and the results are always some variation on what you see here.

Any ideas how I might further debug this, and hopefully get it working?

**Comment by craigshelley** [comment]
On Sun, 2015-12-27 at 20:00 -0800, greedy9 wrote:
> Trying to decode pager transmissions, using rtl_fm and multimon-ng on
> a Raspberry Pi This is what I've assembled from various sources, and
> it almost works
> 
> rtl_fm -A fast -f 929620M -M fm -E dc -s 22050 -F 0 | multimon-ng -n
> -p -t raw -a POCSAG512 -a POCSAG1200 -a POCSAG2400 -a FLEX -i -p -v 3
> -f alpha /dev/stdin
> 
> I see messages from the FLEX decoder, mostly these:
> 
> FLEX: Synchronisation Lost
> FLEX: Locked
> FLEX: Synchronisation Lost
> FLEX: Locked
> 
> but periodically, I get:
> 
> FLEX: Locked
> FLEX: SyncInfoWord: sync_code=0xd6a0 baud=3200 levels=4 polarity=NEG
> zero=0013193 envelope=0195820 symrate=1601396648
> FLEX: State: FIW
> FLEX: Synchronisation Lost
> FLEX: State: SYNC1
> FLEX: Locked
> FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=NEG
> zero=0017672 envelope=0191791 symrate=1600031621
> FLEX: State: FIW
> FLEX: Phase F Data corruption - Unable to fix errors
> FLEX: Unable to decode FIW, too much data corruption
> FLEX: State: SYNC1
> FLEX: Synchronisation Lost
> 
> and a few of these:
> 
> FLEX: State: FIW
> FLEX: Phase F Fixed 2 errors @ 0x00084000 (0x7ffef99e -> 0x7ff6b99e)
> FLEX: Bad Checksum 0x2
> 
> which seems to suggest that multimon-ng is detecting the FLEX format,
> and decoding some of it, but never completely successfully 
> 
> I've played with the gain, antennas, and so on, and the results are
> always some variation on what you see here
> 
> Any ideas how I might further debug this, and hopefully get it
> working?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> 
Hi Greedy9,

Please could you post or provide a link to an audio sample. I can't
guarantee that I'll be able to look into it, but at least I'll be able
to recreate the problem you are observing.

Regards,
Craig

**Comment by greedy9** [comment]
Hi Craig:

Great.  I'll be glad for any help or advice.

Wav file is attached.  The squelch was open, so there is receiver noise between the (short) data bursts.

It was captured with this command:

rtl_fm -d 1 -A fast -f 929.620M -M fm -E dc -s 22050 -g 50 -F 0 | sox -t raw -e signed -c 1 -b 16 -r 22050 - flex2.wav 

Had to zip the file to satisfy the 
[flex2.wav.zip](https://github.com/EliasOenal/multimon-ng/files/76130/flex2.wav.zip)
 GitHub uploader.

Thanks,

Jerry

**Comment by j-dickinson** [comment]
I'm having almost exactly the same issue, although my raw data is coming via UDP packets from `gqrx` through `nc` and `sox`, like this:

```
nc -l -u -p 5000 | sox -t raw -esigned-integer -b 16 -r 48000 - -esigned-integer -b 15 -r 22050 -t raw - | multimon-ng -t raw -a POCSAG512 -a POCSAG1200 -a POCSAG2400  -a FLEX -v 3 -f alpha -
```

Here's some sample output:

```
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=POS zero=-0.037334 envelope=0.124534 symrate=1599.906977
FLEX: State: FIW
FLEX: Phase F Data corruption - Unable to fix errors.
FLEX: Unable to decode FIW, too much data corruption
FLEX: State: SYNC1
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=POS zero=-0.037923 envelope=0.124479 symrate=1600.049702
FLEX: State: FIW
FLEX: Phase F Fixed 2 errors @ 0x00000012  (0x30bb1071 -> 0x30bb1063)
FLEX: Bad Checksum 0x3
FLEX: State: SYNC1
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=POS zero=-0.037626 envelope=0.124186 symrate=1600.049702
FLEX: State: FIW
FLEX: Phase F Fixed 2 errors @ 0x00200200  (0x7c335879 -> 0x7c135a79)
FLEX: Bad Checksum 0x7
FLEX: State: SYNC1
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: Synchronisation Lost
FLEX: Locked
```

This setup *does* work for some FLEX channels, but not others. When I dug into it, I noticed that the failing ones were sending at 3200 baud 4 levels positive polarity, while the working ones were sending at 1600 baud 4 levels positive polarity.

Maybe there is a timing error in either my setup or multimon-ng which is only a showstopping error when the faster baud rate is in play?

**Comment by j-dickinson** [comment]
Ok, I think I found the problem. At the end of SYNC1, in `decode_mode`, the decoder finds out that the baud rate of the pager packet is going to be 3200 and immediately sets `flex->Sync.baud` to 3200. This causes it to start decoding all subsequent symbols at 3200 baud. However, this is not correct - the FIW is sent at 1600 baud, so we should continue to decode at 1600 baud until the FIW is done.

This problem is masked for 1600 baud packets because SYNC2 and SYNC1/FIW are at the same rate.

I can't immediately find a clean way to resolve this in the framework of the current code, but I'll take a look a little later if somebody else doesn't get to it first.

**Comment by j-dickinson** [comment]
There were actually a few places where the code was not correct for 3200 baud. Looking back at the original pull request for the decoder, #33, this makes sense:

> 3200 symbols per second is supported, however it has not been tested.

I think I've fixed them all in my local checkout, but I'd like to test it some more before committing. I'll then create a pull request.

**Comment by j-dickinson** [comment]
Note that despite fixing FLEX decoding for me, the fix in #47 only decodes one of the two FLEX messages in @greedy9's sample above:

```
$ cat ~/Downloads/flex2.wav | multimon-ng -t raw -a POCSAG512 -a POCSAG1200 -a POCSAG2400  -a FLEX -a SCOPE -v 3 -f alpha -
multimon-ng  (C) 1996/1997 by Tom Sailer HB9JNX/AE4WA
             (C) 2012-2014 by Elias Oenal
available demodulators: POCSAG512 POCSAG1200 POCSAG2400 FLEX EAS UFSK1200 CLIPFSK FMSFSK AFSK1200 AFSK2400 AFSK2400_2 AFSK2400_3 HAPN4800 FSK9600 DTMF ZVEI1 ZVEI2 ZVEI3 DZVEI PZVEI EEA EIA CCIR MORSE_CW DUMPCSV SCOPE
Enabled demodulators: POCSAG512 POCSAG1200 POCSAG2400 FLEX SCOPE
FLEX: Locked
FLEX: Synchronisation Lost
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=POS zero=0.074785 envelope=0.206712 symrate=1598.460775
FLEX: State: FIW
FLEX: FrameInfoWord: cycleno=05 frameno=033 fix3=0x38 time=21:01
FLEX: State: SYNC2
FLEX: State: DATA
FLEX: Synchronisation Lost
FLEX: State: SYNC1
FLEX: Locked
FLEX: SyncInfoWord: sync_code=0xdea0 baud=3200 levels=4 polarity=POS zero=0.058008 envelope=0.205656 symrate=1600.049702
FLEX: State: FIW
FLEX: FrameInfoWord: cycleno=05 frameno=034 fix3=0x3c time=21:03
FLEX: State: SYNC2
FLEX: State: DATA
FLEX: BlockInfoWord: (Phase A) BIW:0000040B AW:01-01 (0 pages)
FLEX: BlockInfoWord: (Phase B) BIW:0000040B AW:01-01 (0 pages)
FLEX: BlockInfoWord: (Phase C) BIW:0000040B AW:01-01 (0 pages)
FLEX: BlockInfoWord: (Phase D) BIW:0000040B AW:01-01 (0 pages)
FLEX: State: SYNC1
FLEX: Synchronisation Lost
```

The inability to decode the first message might be a remaining bug in the code, or it might be due to noise in the sample.
